[DO NOT MERGE] Selection Components FAQ (#6999)
elastic/eui: wiki/temp/selection-components.md

# The Ultimate Guide to EUI Selection Components

EUI offers different selection solutions and choosing the right one  may seem tricky. Welcome to a comprehensive guide on the EUI selection components. This guide will include explanations and comparisons for each component. 

> ![Spiderman Selection components meme](spiderman_selection.jpeg)

Reviewer: I love this meme! It's such a fun way to communicate the similarities and differences in our components. Because it's such an important part of this page's context, could we expand on the alt text? Maybe a short description of multiple Spidermen pointing at each other, and an explanation of how that relates to our selection components.

```
Five Spidermen are in a room, pointing at one another, as if to say "It's you! No, it's you!" Each Spiderman has the name of an EUI selection component next to them.
```

Author: The first time I read this, my eyes saw "five Spidermen walk into a bar". The beginnings of a corny joke 😆

## Meet the EUI selection components

EUI currently offers **5** different selection solutions. 

### [`EuiSelect`](https://elastic.github.io/eui/#/forms/form-controls#select)

> EuiSelect renders a basic HTML `<select>` element. It should be used when you need to render 7 to 12 single items where only one option will be selected. This is a no-frills selection component that renders options as a string. If you would like to customize how the options are rendered and selected, you can use [`EuiSuperSelect`](#euisuperselect) instead. For  lists over 12 items, consider using [`EuiComboBox`](#euicombobox) which has search and multi-select capabilities. 

Reviewer 2: these lines are kind of pretty long and a nightmare to code suggest on so just going to add this feedback - `7 to 12 single items` can just be `7 to 12 items`, the **selection** is single, not the items/options (which are always technically single? - so there isn't a need to describe them as such)

> EuiSelect renders a basic HTML `<select>` element. It should be used when you need to render 7 to 12 single items where only one option will be selected. This is a no-frills selection component that renders options as a string. If you would like to customize how the options are rendered and selected, you can use [`EuiSuperSelect`](#euisuperselect) instead. For  lists over 12 items, consider using [`EuiComboBox`](#euicombobox) which has search and multi-select capabilities. 

Reviewer 2: Also mentioning as a v minor nit, `For  lists` has two spaces instead of just one here - I've noticed one or two other double spaces like this in the file so you may want to do a `ctrl+f` across the board for it. (also I know it doesn't affect rendered output so I'm totally being a butt here, however my inner copyeditor cannot be quelled 👹)

> EuiSelect renders a basic HTML `<select>` element. It should be used when you need to render 7 to 12 single items where only one option will be selected. This is a no-frills selection component that renders options as a string. If you would like to customize how the options are rendered and selected, you can use [`EuiSuperSelect`](#euisuperselect) instead. For  lists over 12 items, consider using [`EuiComboBox`](#euicombobox) which has search and multi-select capabilities. 

Reviewer 2: Alright, actual content based feedback:

Let's add a paragraph break between your first sentence and your second. Your first one is describing the component; the second is describing when _not_ to use the component. I would strongly rather separate the two if possible, and add a bit more detail to the first:

> [first sentence w/ basic description of EuiSelect as a native select component]. **In general,** we strongly recommend reaching for `EuiSelect` first when possible over other more complex selection components. Native form controls tend to have better cross-browser and cross-device affordances (particularly for mobile devices), and are more familiar to end users due to their ubiquity across the internet.
>
> If you have to customize selection display or selection, [...rest of second sentence]

### [`EuiSuperSelect`](https://elastic.github.io/eui/#/forms/super-select)

> `EuiSuperSelect` gives you more control over the display of the options in the dropdown. The `dropdownDisplay prop` gives you the ability to pass in full React nodes for each option, allowing for multi-line option text or option descriptions.

Reviewer 2: `full React nodes` -> `custom React nodes, just IMO

### [`EuiComboBox`](https://elastic.github.io/eui/#/forms/combo-box)

> `EuiComboBox` allows multiple items to be selected and shown all at once. This should be used when the user has a number of options to choose from and needs to be able to search for them. `EuiComboBox` also allows users to specify a custom value in addition to selecting from a predetermined list.

Reviewer: I took a pass at this last sentence using Hemingway. What do you think about revising it to an "or" statement that highlights both abilities?

```suggestion
`EuiComboBox` allows multiple items to be selected and shown all at once. This should be used when the user has a number of options to choose from, and needs to be able to search for them. `EuiComboBox` allows users to specify a custom value or select from a rendered list.
```

Reviewer 2: > This should be used when the user has a number of options to choose from and needs to be able to search for them.

TBH, I disagree with this statement. While users **can** search to filter dropdown options, they can also simply select from the dropdown menu as well. IMO it's not a "needs" situation, in the same way that using EuiSelectable with a searchbox is a clear consumer choice based on their knowledge of the number of options. I would say what most **clearly delineates** EuiComboBox from EuiSelectable is 1. pill display, and 2. **the ability to support custom user-entered values**. Otherwise EuiComboBox and EuiSelectable both the same features of searching and selecting multiple options.

### [`EuiSelectable`](https://elastic.github.io/eui/#/forms/selectable)
`EuiSelectable` renders a searchable list where multiple options can be selected or excluded. It should not be used for primary navigation, but can be used to simplify the construction of popover navigational menus. `EuiSelectable` does not appear in a popover or show all selected items in a list.
 
 #### [`EuiSelectableTemplateSitewide`](https://elastic.github.io/eui/#/templates/sitewide-search)

>  `EuiSelectableTemplateSitewide` is an opinionated wrapper around `EuiSelectable`.  It creates the search input that triggers a popover containing a list of navigation options. Options many include a  display label, icon, avatar, and meta text.

Reviewer 2: I'd like to see us include more information as to **when** to use certain components as opposed to simply describing them. e.g.,

```suggestion
`EuiSelectableTemplateSitewide` is an opinionated wrapper around `EuiSelectable` and is an example of the flexibility that `EuiSelectable` offers. It renders a search input that triggers a popover containing a list of navigation options. Options many include a display label, icon, avatar, and meta text.
 
In general, we recommend its usage for a global search bar from an app-wide page header.
```

### [`EuiSuggest`](https://elastic.github.io/eui/#/forms/suggest)
`EuiSuggest` is a text field used to display suggestions. The status of each suggestion is displays on its right side. Statuses include `saved`, `unsaved`,  `unchanged`, and `isLoading`. Include `type` and `label` in your list of suggestions to customize each option. 

## Comparing selection components

Below is a handy reference for comparing the selection components.

|   | `EuiSelect` | `EuiSuperSelect` | `EuiSelectable` | `EuiComboBox` | `EuiSuggest` |
|---|:---:|:---:|:---:|:---:|:---:|
| Select a single option  | ✅ | ✅ | ✅ | ✅ | ✅ |
| Select multiple options  | ❌ | ❌ | ✅ | ✅ | ❌ |
| Accepts custom values from users | ❌ | ❌ | ❌ | ✅  | ❌ |
| Option exclusion | ❌ | ❌ | ✅ | ❌ | ❌ |
| Customizable option display | ❌ | ✅ | ❌ | ✅ | ❌
| Controllable loading state | ✅ | ✅ | ✅ | ✅ | ✅ |

> | Customizable loading state | ✅ | ✅ | ❌ | ✅ | ✅ |

Reviewer 2: I'm confused about what the difference bis etween controllable vs customizable loading state.

I also don't think the table is correct for customizable loading state, as of all the components, EuiSelectable **does** have a customizable `loadingMessage` whereas the other components only have a basic `isLoading` bool.

Either way, I'm not sure this is totally worth flagging, or I'd combine 'Customizable loading state' with 'Customizable error handling' down below to just a single `Customizable loading/error messaging`

Author: Bree's Brain: 
`Controllable Loading State` -> Using the `isLoading` prop to allow consumers to decide when the component is loading.
`Customizable Loading State` -> Creating custom functionality when it is loading. In this case, a it would be a custom loading message. `EuiSelectable` is the only one of the bunch that offers this. 

I'll get rid of this row and merge the loading/error messaging!

| Customizable error handling | ❌ | ❌ | ✅ | ❌ | ❌ 

> | Searchable | ❌ | ❌ | ✅ | ✅ | ❌ |

Reviewer 2: This is incorrect - `EuiSuggest` dogfoods `EuiSelectable` and is 100% searchable.

Also, I don't feel EuiSelect deserves a solid red X - while it isn't **technical** searchable, it still follows native select rules, which means if you start to type in the option you want, the browser will automatically try and jump you to the option that matches what you're typing. (which is why I love native selects and get real annoyed at websites that use custom selects that don't have this behavior, especially for date/years of birth 😄)

| Accepts custom values from users | ❌ | ❌ | ❌ | ✅ | ❌ |

> | `disabled` state | ✅ <br/> (Disables whole component) | ✅ <br/> (Disables individual option) | ✅ <br/> (Disables individual option) | ✅ <br/> (Disables whole component) | ❌ |

Reviewer 2: I really like this nuance describing disabled state, nice work!

lmfao @ EuiSuggest not supporting `disabled` state though, that seems like such an oversight 💀 we seriously need to get rid of this component

Reviewer 2: Haha so I just searched and there's 0 usages of EuiSuggest in Kibana. Zero! My desire to remove it just shot up tenfold

> | `readOnly` state | ❌ | ✅ | ❌ | ❌ | ❌ |

Reviewer 2: TIL [native select elements don't support `readOnly`](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/readonly). At first I was like 'this seems like a bug' but now I'm like whoaaaa

> | Built in utility function for clearing user input | ❌ | ❌ | ✅ | ✅  | ✅ 

Reviewer: I ran the page through VoiceOver to see how it handled the icons. They read back as "Checkmark" and "Exe" which offer pretty good semantic clues this component does or does not offer this feature. I tried out adding "Yes" and "No" text to a few cells. It did add clarity for screen reader use, so I'll float it as an enhancement idea. Would look roughly like this sketch in code:

<img width="930" alt="Screen Shot 2023-07-27 at 8 42 14 AM" src="https://github.com/elastic/eui/assets/934879/f3a4c490-5c1e-40c8-812f-db05aae7e059">

> | Built in utility function for clearing user input | ❌ | ❌ | ✅ | ✅  | ✅ 

Reviewer 2: > Built in utility function for clearing user input

The table documentation for this is not quite correct or consistent:

- EuiSelectable allows clearing **the current search value** but not clearing all selections
- EuiComboBox actually allows clearing all selections
- EuiSuggest is also only clearing the search value (it has no concept of internal selection state and thus cannot clear selections)

You could separate this into "allows clearing search" vs "allows clearing selection", but I'm not sure how useful that would be. "Built in utility for clearing all selections" would probably be the most useful feature people want, and that would only go to EuiComboBox (although I know people have requested it for EuiSelectable).

---

> Still have a question? Feel free to comment below, reach out to us via our Slack channel (#eui), or open an issue. We're always happy to help!

Reviewer: I think having this section at the end would benefit from a heading, probably an H2. I added a link to the open an issue text so community users have a direct line to action.

```suggestion
## Still have a question? ##
Feel free to comment below, reach out to us via our Slack channel (#eui), or [open an issue](https://github.com/elastic/eui/issues/new/choose). We're always happy to help!
```

### [`EuiSelect`](https://elastic.github.io/eui/#/forms/form-controls#select)
`EuiSelect` renders a basic HTML `<select>` element. This is a no-frills selection component that renders `<options>` as a string. It should be used when you need to render 7 to 12 items where only one option will be selected.

In general, we strongly recommend reaching for `EuiSelect` first when possible over other more complex selection components. Native form controls tend to have better cross-browser and cross-device affordances (particularly for mobile devices), and are more familiar to end users due to their ubiquity across the internet.

If you need to customize how the options are rendered and selected, you can use [EuiSuperSelect](#euisuperselect) instead. For lists over 12 items, consider using EuiSelectable which has search and multi-select capabilities.

### [`EuiSuperSelect`](https://elastic.github.io/eui/#/forms/super-select)
`EuiSuperSelect` gives you more control over the display of the options in the dropdown. The `dropdownDisplay` prop gives you the ability to pass in custom React nodes for each option, allowing for multi-line option text or descriptions.

> In general, we recommend reaching for `EuiSuperSelect` only if custom option display is absolutely necessary [see EuiSelect’s paragraph on accessibility and UX](https://elastic.github.io/eui/#/forms/form-controls#select).

Reviewer 2: This link doesn't go anywhere - were you planning on adding that as part of this work? or should we just go ahead and remove that URL and sentence.
```suggestion
In general, we recommend reaching for `EuiSuperSelect` only if custom option display is absolutely necessary, as `EuiSelect` is natively more accessible.
```

### [`EuiComboBox`](https://elastic.github.io/eui/#/forms/combo-box)

> `EuiComboBox` allows multiple items to be selected and shown in a horizontally-oriented pill list. It allows users to search through the list of dropdown options, and additionally supports custom user-generated input outside of the available options.

Reviewer 2: minor grammar suggestion
```suggestion
`EuiComboBox` allows multiple items to be selected and shown in a horizontally-oriented pill list. It allows users to search through a list of dropdown options, and additionally supports custom user-generated input outside of the available options.
```
